File: app/src/main/java/dev/ragnarok/fenrir/upload/impl/StoryUploadable.java
package dev.ragnarok.fenrir.upload.impl;

import static dev.ragnarok.fenrir.util.RxUtils.safelyCloseAction;
import static dev.ragnarok.fenrir.util.Utils.listEmptyIfNull;
import static dev.ragnarok.fenrir.util.Utils.safelyClose;

import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.provider.MediaStore;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.File;
import java.io.FileInputStream;
import java.io.InputStream;
import java.util.Collections;

import dev.ragnarok.fenrir.api.PercentagePublisher;
import dev.ragnarok.fenrir.api.interfaces.INetworker;
import dev.ragnarok.fenrir.api.model.server.UploadServer;
import dev.ragnarok.fenrir.domain.IOwnersRepository;
import dev.ragnarok.fenrir.domain.Repository;
import dev.ragnarok.fenrir.domain.mappers.Dto2Model;
import dev.ragnarok.fenrir.exception.NotFoundException;
import dev.ragnarok.fenrir.model.Story;
import dev.ragnarok.fenrir.settings.Settings;
import dev.ragnarok.fenrir.upload.IUploadable;
import dev.ragnarok.fenrir.upload.MessageMethod;
import dev.ragnarok.fenrir.upload.Upload;
import dev.ragnarok.fenrir.upload.UploadResult;
import io.reactivex.rxjava3.core.Single;

public class StoryUploadable implements IUploadable<Story> {

    private final Context context;
    private final INetworker networker;

    public StoryUploadable(Context context, INetworker networker) {
        this.context = context;
        this.networker = networker;
    }

    private static String findFileName(Context context, Uri uri) {
        String fileName = uri.getLastPathSegment();
        try {
            String scheme = uri.getScheme();
            if (scheme.equals("file")) {
                fileName = uri.getLastPathSegment();
            } else if (scheme.equals("content")) {
                String[] proj = {MediaStore.MediaColumns.DISPLAY_NAME};

                Cursor cursor = context.getContentResolver().query(uri, proj, null, null, null);
                if (cursor != null && cursor.getCount() != 0) {
                    int columnIndex = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.DISPLAY_NAME);
                    cursor.moveToFirst();
                    fileName = cursor.getString(columnIndex);
                }

                if (cursor != null) {
                    cursor.close();
                }
            }

        } catch (Exception ignored) {

        }

        return fileName;
    }

    @Override
    public Single<UploadResult<Story>> doUpload(@NonNull Upload upload, @Nullable UploadServer initialServer, @Nullable PercentagePublisher listener) {
        int accountId = upload.getAccountId();

        Single<UploadServer> serverSingle;
        if (initialServer == null) {
            serverSingle = upload.getDestination().getMessageMethod() == MessageMethod.VIDEO ? networker.vkDefault(accountId).users().stories_getVideoUploadServer().map(s -> s) : networker.vkDefault(accountId).users().stories_getPhotoUploadServer().map(s -> s);
        } else {
            serverSingle = Single.just(initialServer);
        }

        return serverSingle.flatMap(server -> {
            InputStream[] is = new InputStream[1];

            try {
                Uri uri = upload.getFileUri();

                File file = new File(uri.getPath());
                if (file.isFile()) {
                    is[0] = new FileInputStream(file);
                } else {
                    is[0] = context.getContentResolver().openInputStream(uri);
                }

                if (is[0] == null) {
                    return Single.error(new NotFoundException("Unable to open InputStream, URI: " + uri));
                }

                String filename = findFileName(context, uri);
                return networker.uploads()
                        .uploadStoryRx(server.getUrl(), filename, is[0], listener, upload.getDestination().getMessageMethod() == MessageMethod.VIDEO)
                        .doFinally(safelyCloseAction(is[0]))
                        .flatMap(dto -> networker
                                .vkDefault(accountId)
                                .users()
                                .stories_save(dto.response.upload_result)
                                .map(items -> listEmptyIfNull(items.getItems()))
                                .flatMap(tmpList -> Repository.INSTANCE.getOwners().findBaseOwnersDataAsBundle(accountId, Collections.singletonList(Settings.get().accounts().getCurrent()), IOwnersRepository.MODE_ANY, null)
                                        .flatMap(owners1 -> {
                                            Story document = Dto2Model.transformStory(tmpList.get(0), owners1);
                                            UploadResult<Story> result = new UploadResult<>(server, document);
                                            return Single.just(result);
                                        })));
            } catch (Exception e) {
                safelyClose(is[0]);
                return Single.error(e);
            }
        });
    }
}

File: app/src/main/java/dev/ragnarok/fenrir/upload/impl/StoryUploadable.kt
package dev.ragnarok.fenrir.upload.impl

import android.content.Context
import dev.ragnarok.fenrir.api.PercentagePublisher
import dev.ragnarok.fenrir.api.interfaces.INetworker
import dev.ragnarok.fenrir.api.model.VKApiStory
import dev.ragnarok.fenrir.api.model.server.UploadServer
import dev.ragnarok.fenrir.domain.IOwnersRepository
import dev.ragnarok.fenrir.domain.Repository.owners
import dev.ragnarok.fenrir.domain.mappers.Dto2Model
import dev.ragnarok.fenrir.exception.NotFoundException
import dev.ragnarok.fenrir.model.Story
import dev.ragnarok.fenrir.settings.Settings
import dev.ragnarok.fenrir.upload.*
import dev.ragnarok.fenrir.util.RxUtils.safelyCloseAction
import dev.ragnarok.fenrir.util.Utils.listEmptyIfNull
import dev.ragnarok.fenrir.util.Utils.safelyClose
import io.reactivex.rxjava3.core.Single
import java.io.File
import java.io.FileInputStream
import java.io.InputStream

class StoryUploadable(private val context: Context, private val networker: INetworker) :
    IUploadable<Story> {
    override fun doUpload(
        upload: Upload,
        initialServer: UploadServer?,
        listener: PercentagePublisher?
    ): Single<UploadResult<Story>> {
        val accountId = upload.accountId
        val serverSingle: Single<UploadServer> = if (initialServer == null) {
            if (upload.destination.messageMethod == MessageMethod.VIDEO) networker.vkDefault(
                accountId
            ).users().stories_getVideoUploadServer()
                .map { it } else networker.vkDefault(accountId).users()
                .stories_getPhotoUploadServer().map { it }
        } else {
            Single.just(initialServer)
        }
        return serverSingle.flatMap { server ->
            val `is` = arrayOfNulls<InputStream>(1)
            try {
                val uri = upload.fileUri
                val file = File(uri!!.path!!)
                if (file.isFile) {
                    `is`[0] = FileInputStream(file)
                } else {
                    `is`[0] = context.contentResolver.openInputStream(uri)
                }
                if (`is`[0] == null) {
                    return@flatMap Single.error<UploadResult<Story>>(
                        NotFoundException(
                            "Unable to open InputStream, URI: $uri"
                        )
                    )
                }
                val filename = UploadUtils.findFileName(context, uri)
                return@flatMap networker.uploads()
                    .uploadStoryRx(
                        server.url,
                        filename,
                        `is`[0]!!,
                        listener,
                        upload.destination.messageMethod == MessageMethod.VIDEO
                    )
                    .doFinally(safelyCloseAction(`is`[0]))
                    .flatMap { dto ->
                        networker
                            .vkDefault(accountId)
                            .users()
                            .stories_save(dto.response.upload_result)
                            .map {
                                listEmptyIfNull<VKApiStory>(it.getItems())
                            }
                            .flatMap { tmpList ->
                                owners.findBaseOwnersDataAsBundle(
                                    accountId, listOf(
                                        Settings.get().accounts().current
                                    ), IOwnersRepository.MODE_ANY, null
                                )
                                    .flatMap {
                                        val document =
                                            Dto2Model.transformStory(tmpList[0], it)
                                        val result = UploadResult(server, document)
                                        Single.just(result)
                                    }
                            }
                    }
            } catch (e: Exception) {
                safelyClose(`is`[0])
                Single.error(e)
            }
        }
    }
}